slogr/slogr.go

```diff
@@ -0,0 +1,46 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package slogr
+
+import (
+	"log/slog"
+
+	"github.com/go-logr/logr"
+)
+
+// New returns a logr.Logger which writes to the same handler as the slog.Logger.
+//
+// In the output the logr verbosity level gets negated, so V(4) becomes
+// slog.LevelDebug.
+func New(logger *slog.Logger) logr.Logger {
+	return logr.New(&slogSink{handler: logger.Handler()})
+}
+
+// NewSlogLogger returns a slog.Logger which writes to the same sink as the logr.Logger.
+//
+// The returned logger writes all records with level >= slog.LevelError as
+// error log entries with LogSink.Error, regardless of the verbosity of the
+// logr.Logger. The level of all other records gets reduced by the verbosity
+// level of the logr.Logger, so a slog.Logger.Info call gets written with
+// slog.LevelDebug when using a logr.Logger where verbosity was modified with
+// V(4).
+func NewSlogLogger(logger logr.Logger) *slog.Logger {
+	return slog.New(&slogHandler{sink: logger.GetSink(), level: slog.Level(logger.GetV())})
```

Comment:
Returning a `slog.Logger` is done here for consistency with the return code of `New` above ("return high-level API"). We concluded in https://github.com/go-logr/logr/pull/196#issuecomment-1665978463 that `slog.Logger` is what we want to support initially.

But `New` above doesn't do anything with the `slog.Logger`, it could also accept a `slog.Handler`. Instead of `NewSlogLogger`, users could do `slog.New(NewSlogHandler(...))`. That's then consistent with e.g. `slog.NewTextHandler`. 

Overall I am therefore leaning towards using `slog.Handler` as the slog API that slogr accepts and returns.

Reply:
I think I agree about Handler.  Less sure about names.

NewLogrHandler would be the consistent name, but confusing and sort of pointless.

There a bit of asymmetry - logr implementations (zapr, funcr) call logr.New and return the Logger.  Slog implementations return a Handler and expect the caller to call slog.New.  I don't think we should fight that.  If there's a need to produce a logr.LogSink, that would be distinct from returning a Logger. For slog, I think Handler is the common unit.

Reply:
I agree.

Reply:
Done.